### DISCUSSION TITLE
Why is the debug apk significantly bigger compared to cargo-apk

### DISCUSSION THREAD

**Discussion by NiklasEi** [rust-mobile/xbuild#170] [Q&A] [answered]
With `cargo-apk`, my debug apk was ~70Mb. With `xbuild` it's ten times as large (~750Mb). This is not a deal breaker, since the release builds seem comparable, but I am wondering where the apk size differences are coming from. 

Maybe related: why did I have to install new dependencies to get the app to compile ("clang", "lld", "llvm") when `cargo-apk` was able to compile a working app without them?

I had a very similar size with `cargo-apk` before striping debug symbols from the shared libc++ library, but looking into the `apk` file, the included `libc++_shared.so` is only 7Mb while my "main" library is 700Mb large. :thinking:

**Comment by dvc94ch** [comment, marked as answer]
When you unzip the apk, what is the large part?

**Reply by NiklasEi** [reply]
My own library (`.so`).

**Reply by dvc94ch** [reply]
the library is built by cargo, it's not obvious why one would be so much larger than the other. can you use nm and readelf to explore the two libraries to figure out why one is so much bigger than the other? Using readelf to show the section sizes should give you a first idea of which sections are present and are abnormally larger than expected.

**Reply by NiklasEi** [reply]
Almost the complete size is debug sections:
```
libbrain_games_mobile.so  :
section                    size       addr
.note.android.ident         152        624
.note.gnu.build-id           24        776
.dynsym                    5784        800
.gnu.version                482       6584
.gnu.version_r              128       7068
.gnu.hash                   388       7200
.hash                      1936       7588
.dynstr                    5402       9524
.rela.dyn               2905440      14928
.rela.plt                  5400    2920368
.rodata                 2160456    2925824
.gcc_except_table       1156632    5086280
.eh_frame_hdr            578804    6242912
.eh_frame               3044932    6821720
.text                  18454040    9870748
.plt                       3632   28324800
.data.rel.ro            1694648   28332528
.fini_array                  16   30027176
.dynamic                    464   30027192
.got                       5792   30027656
.got.plt                   1824   30033448
.data                      5616   30039368
.bss                       7128   30044984
.comment                    221          0
.debug_loc             95515247          0
.debug_abbrev           2064565          0
.debug_info           163873574          0
.debug_aranges          2271984          0
.debug_ranges          55896720          0
.debug_str            122030404          0
.debug_pubnames        93757624          0
.debug_pubtypes        89564629          0
.debug_line            26490498          0
Total                 681504586
```

The debug apk build with `cargo-apk` compiling the same code with the same `Cargo.toml` does not have these debug sections. But this makes sense since I configured `strip = "strip"` for `cargo-apk`. Thank you for your help!

**Reply by MarijnS95** [reply]
> But this makes sense since I configured `strip = "strip"` for `cargo-apk`

Oh, oof :grinning:

**Reply by NiklasEi** [reply]
Yeah, for some reason I did not think of my game as a shared library and though that this option only removes the debug symbols from `libc++_shared.so` :see_no_evil:

**Comment by dvc94ch** [comment]
The reason is because cargo-apk downloads the prebuilt android toolchain, but we use the toolchain distributed with the os. Using the special android toolchain means that we need a lot of code specific to android, the current setup works for all platforms.
